### ISSUE TITLE
Refactor UCEF3 package

### ISSUE THREAD

**Issue by pmpowers-usgs** [opensha/opensha-ucerf3#3]
Refactor UCERF3 package (from `scratch`) to ?? org.opensha.ucerf3

We should give some thought as to how we want ucerf3 organized. Things like FaultSystemSolution were borne out of UCERF3 development but aren't specific to the model itself. Same goes for the inversion methodology and model analysis and verification tools.

**Comment by kevinmilner** [comment]
Agreed, especially with respect to FaultSystem[Solution/RupSet]. It should probably head to core. InversionFaultSystem[Solution/RupSet] would stay in ucerf3.

**Comment by kevinmilner** [comment]
I think the time has come to finally do this, as we are moving toward UCERF4 and other regions are now using UCERF3 inversion code (other US states, NZ). The change will be really tedious and will surely bring up issues not discussed below, but it won't get any easier if we continue to put it off.

Here is my current thinking (this list is not exhaustive, only major classes mentioned):

## opensha-core

The following classes will be promoted from the current UCERF3 project, and any UCERF3-specific parts would be moved to U3-specific subclasses in the UCERF repository.

* org.opensha.sha.earthquake.faultSysSolution
  * FaultSystemRupSet
  * FaultSystemSolution
  * SlipEnabledRupSet (this adds average slip in rupture information to fault system rup sets)
  * SlipEnabledSolution
* org.opensha.sha.earthquake.faultSysSolution.utils
  * FaultSystemIO
  * MatrixIO
* org.opensha.sha.earthquake.faultSysSolution.griddedSeismicity
  * GridSourceProvider
* org.opensha.sha.earthquake.faultSysSolution.plausibility (renamed from "laugh test")
  * PlausibilityFilter
  * AbstractPlausibilityFilter
* org.opensha.sha.earthquake.faultSysSolution.plausibility.impl
  * (most of the plausibility filters here, but maybe not coulomb?)
* org.opensha.sha.calc.simulatedAnnealing
  * SimulatedAnnealing
  * SerialSimulatedAnnealing
  * ThreadedSimulatedAnnealing
  * 'completion' and 'params' sub-packages
* org.opensha.sha.util.logicTree
  * *most things from scratch.UCERF3.logicTree, but moving the UCERF3-specific static references to a new U3LogicTreeBranch class*

Notes:
* This will be a bit tricky in that FaultSystemIO currently will load/write a FSS zip file as an InversionFaultSystemSolution if applicable. If IFSS lives in another project, then there will have to be a separate downstream I/O class, still thinking about this. I'll investigate generalizing IFSS as well.

## opensha-ucerf

* rename 'opensha-ucerf3' repository to opensha-ucerf
  * this will be the home for UCERF4+ and UCERF3
  * this should be [transparent to existing clones](https://docs.github.com/en/github/administering-a-repository/renaming-a-repository), but might get tricky with gradle
* create org.opensha.ucerf and move things likely to be common to UCERF3 and UCERF4 here
* move everything else from scratch.UCERF3 to org.opensha.ucerf.ucerf3 (or ...ucerf.v3? thoughts?)
* keep FaultSystemSolutionERF here for now as it contains UCERF3-specific probability models, and potentially separate out a more generic parent class in the future.

**Comment by field-usgs** [comment]
Good idea Kevin, and sorry if I’m not replying in the best way.

I’m ccing Peter because this relates to the US NSHM, and we all need to decide (once and for all) what happening with our respective codes.

I think "opensha-ucerf” is too CA specific.  What about “opensha-fsserf” where fss is for fault-system solution.  Or maybe we just use opensha-fserf (leave out "solution”)?




On Jul 14, 2020, at 7:21 PM, Kevin Milner <notifications@github.com<mailto:notifications@github.com>> wrote:


I think the time has come to finally do this, as we are moving toward UCERF4 and other regions are now using UCERF3 inversion code (other US states, NZ). The change will be really tedious and will surely bring up issues not discussed below, but it won't get any easier if we continue to put it off.

Here is my current thinking (this list is not exhaustive, only major classes mentioned):

opensha-core

The following classes will be promoted from the current UCERF3 project, and any UCERF3-specific parts would be moved to U3-specific subclasses in the UCERF repository.

  *   org.opensha.sha.earthquake.faultSysSolution
     *   FaultSystemRupSet
     *   FaultSystemSolution
     *   SlipEnabledRupSet (this adds average slip in rupture information to fault system rup sets)
     *   SlipEnabledSolution
  *   org.opensha.sha.earthquake.faultSysSolution.utils
     *   FaultSystemIO
     *   MatrixIO
  *   org.opensha.sha.earthquake.faultSysSolution.griddedSeismicity
     *   GridSourceProvider
  *   org.opensha.sha.earthquake.faultSysSolution.plausibility (renamed from "laugh test")
     *   PlausibilityFilter
     *   AbstractPlausibilityFilter
  *   org.opensha.sha.earthquake.faultSysSolution.plausibility.impl
     *   (most of the plausibility filters here, but maybe not coulomb?)
  *   org.opensha.sha.calc.simulatedAnnealing
     *   SimulatedAnnealing
     *   SerialSimulatedAnnealing
     *   ThreadedSimulatedAnnealing
     *   'completion' and 'params' sub-packages
  *   org.opensha.sha.util.logicTree
     *   most things from scratch.UCERF3.logicTree, but moving the UCERF3-specific static references to a new U3LogicTreeBranch class

Notes:

  *   This will be a bit tricky in that FaultSystemIO currently will load/write a FSS zip file as an InversionFaultSystemSolution if applicable. If IFSS lives in another project, then there will have to be a separate downstream I/O class, still thinking about this. I'll investigate generalizing IFSS as well.

opensha-ucerf

  *   rename 'opensha-ucerf3' repository to opensha-ucerf
     *   this will be the home for UCERF4+ and UCERF3
     *   this should be transparent to existing clones<https://docs.github.com/en/github/administering-a-repository/renaming-a-repository>, but might get tricky with gradle
  *   create org.opensha.ucerf and move things likely to be common to UCERF3 and UCERF4 here
  *   move everything else from scratch.UCERF3 to org.opensha.ucerf.ucerf3 (or ...ucerf.v3? thoughts?)
  *   keep FaultSystemSolutionERF here for now as it contains UCERF3-specific probability models, and potentially separate out a more generic parent class in the future.

**Comment by pmpowers-usgs** [comment]
How about `opensha-fault-systems` as an umbrella for the inversion methodology?

**Comment by kevinmilner** [comment]
(Ned, clicking the "view it on GitHub" and replying there is cleanest--Peter and I will be notified of replies and formatting won't get screwed up)

My thinking was that all of the general "fault system" stuff would be promoted to core, and that we would still have a UCERF repository which would store all of the data (what is currently `scratch/UCERF3/data`), UCERF3-specific logic tree branch enums & ERFs (e.g., U3TAS), etc. It would also be the new home of implementation instructions (wiki) and some of the smaller data files. But that does raise the question of where non-CA implementation details would go.

So I think an alternative plan would be to rename `opensha-ucerf3` to `opensha-fault-systems` and to do the class refactoring mentioned above, but not actually promote any of those classes to core. That would certainly be a less painful refactor as well. The downside is that we wouldn't have a clear *this is where UCERF lives* spot, and the UCERF documentation would be intermixed with general FSS documentation (as it currently is on the [to-be-retired Trac site](http://opensha.usc.edu/trac)). I'm fine with this plan if everyone agrees.

**Comment by field-usgs** [comment]
I agree that ideally the base classes would be in core and implementations could be in whatever we want to call it (e.g., opensha-fault-sys-impl).
